File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
import sys
input = sys.stdin.readline


N = int(input())
numbers = list(map(int, input().split()))
ans = [0]

for number in numbers:
    if ans[-1] < number:
        ans.append(number)
    else:
        s = 0
        e = len(ans)
        while s < e:
            mid = (s + e) // 2

            if ans[mid] < number:
                s = mid + 1
            else:
                e = mid
        ans[e] = number

print(len(ans) - 1)

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.util.*;
import java.io.*;

public class Main {
	
	static int[] arr;

	public static void main(String args[]) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		
		arr = new int[N];
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		
		Stack<Integer> stack = new Stack<>();
		stack.push(arr[0]);
		
		for (int i = 1; i < N; i++) {
			int number = arr[i];
			
			if (stack.peek() < number) {
				stack.push(number);
			} else {
				int start = 0;
				int end = stack.size();
				while (start < end) {
					int mid = (start + end) / 2;
					if (stack.get(mid) < number) {
						start = mid + 1;
					} else {
						end = mid;
					}
				}
				stack.set(end, number);
			}
		}
		System.out.println(stack.size());
	}
}